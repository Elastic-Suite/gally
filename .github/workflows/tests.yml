name: Tests

on:
  workflow_call:
    inputs:
      version:
        description: "Version of Gally's packages (ex: master or 1.0.x)"
        type: string
        default: 'master'
        required: false
      last_published_version:
        description: "Last release of the current version"
        required: false
        default: "dev-master"
        type: string
    secrets:
      COMPOSER_AUTH:
        required: true

jobs:
  tests:
    if: github.event.pull_request.draft == false
    name: Tests
    runs-on: ubuntu-latest
    
    env:
      back_branch: ${{ inputs.version }}
      front_branch: ${{ inputs.version == 'master' && 'main' || inputs.version }}
      composer_version: ${{ inputs.version == 'master' && 'dev-master' || (contains(inputs.version, 'x') && format('{0}-dev', inputs.version) || inputs.version) }}
    
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          ref: ${{ env.back_branch }}
      - name: Checkout Gally admin
        uses: actions/checkout@v3
        with:
          repository: 'Elastic-Suite/gally-admin'
          ref: ${{env.front_branch}}
          path: front/gally-admin
      - name: update gally standard package
        working-directory: api
        run: composer require "gally/gally-standard:${{ env.composer_version }} as ${{ inputs.last_published_version }}"
      - name: Add premium package repository
        uses: php-actions/composer@v6
        with:
          php_version: "8.1"
          args: --working-dir api
          command: config repositories.gally-premium composer https://packagist.smile.fr/gally
        env:
          COMPOSER_AUTH: ${{ secrets.COMPOSER_AUTH }}
      - name: Install premium package
        uses: php-actions/composer@v6
        with:
          php_version: "8.1"
          args: --no-cache --working-dir api
          command: require gally/gally-premium:${{ env.composer_version }}
        env:
          COMPOSER_AUTH: ${{ secrets.COMPOSER_AUTH }}
      - name: Pull images
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml pull --ignore-pull-failures || true
      - name: Start services
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml up --build -d --wait
      - name: Check HTTP reachability
        id: http_reachability
        run: curl -v -o /dev/null http://localhost
      - name: HTTP Failure
        if: ${{ failure() && steps.tests.http_reachability == 'failure' }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
      - name: Check HTTPS reachability
        id: https_reachability
        run: curl  -vk -o /dev/null https://localhost
      - name: HTTPS Failure
        if: ${{ failure() && steps.tests.https_reachability == 'failure' }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
      - name: Create test database
        run: |
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console -e test doctrine:database:create
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console -e test doctrine:migrations:migrate --no-interaction
      - name: Doctrine Schema Validator
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console doctrine:schema:validate --skip-sync
      - name: Fixtures load
        run: |
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console gally:index:clear --no-interaction
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console hautelook:fixtures:load --no-interaction -vv
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console doctrine:fixtures:load --append --no-interaction
      - name: Coverage
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=coverage -T php bin/phpunit --coverage-clover=coverage/coverage.xml vendor/gally

      - name: Cleanup coverage file
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -T php sed -i 's:<file name="/srv/:<file name="/:' coverage/coverage.xml

      - if: ${{ github.event_name == 'pull_request' }}
        name: Download artifact
        uses: dawidd6/action-download-artifact@v2.14.1
        continue-on-error: true
        id: base_coverage_report
        with:
          workflow: .github/workflows/ci.yml
          branch: ${{ github.base_ref }}
          name: coverage-report
          path: coverage/base # Retrieve previous coverage file at the root directory, not inside api/coverage

      - name: Backend Coverage Report as Comment (Compared to a base)
        if: ${{ github.event_name == 'pull_request' && steps.base_coverage_report.outcome == 'success' }}
        uses: lucassabreu/comment-coverage-clover@main
        continue-on-error: true
        with:
          with-chart: false
          file: api/coverage/coverage.xml
          base-file: coverage/base/coverage.xml
          max-line-coverage-decrease: 10.0
          max-method-coverage-decrease: 10.0
          min-line-coverage: 75
          min-method-coverage: 75
          signature: "Backend Coverage compared to target branch"

      - name: Backend Coverage Report as Comment (Clean)
        uses: lucassabreu/comment-coverage-clover@v0.3.0
        if: ${{ (github.event_name == 'pull_request' && steps.base_coverage_report.outcome != 'success') || github.event_name == 'push'}}
        with:
          with-chart: false
          file: api/coverage/coverage.xml
          signature: "Backend Coverage of this branch"

      - uses: actions/upload-artifact@v2
        with:
          name: coverage-report
          path: api/coverage/coverage.xml

      - name: Jest
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -T pwa yarn test:ci

      - name: Frontend Coverage Report
        uses: 5monkeys/cobertura-action@v12
        if: ${{ github.event_name == 'pull_request' }}
        with:
          repo_token: "${{ secrets.GITHUB_TOKEN }}"
          report_name: "Frontend Coverage of this branch"
          path: front/example-app/coverage/cobertura-coverage.xml
          minimum_coverage: 1
          fail_below_threshold: false

      - name: "[End] Job failed, gathering logs"
        if: ${{ failure() }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
