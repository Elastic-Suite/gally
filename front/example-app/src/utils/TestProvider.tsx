import { ReactNode } from 'react'

import { IApi, schemaContext } from 'shared'

interface IProps {
  api: IApi
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { api, children } = props
  return <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
}

export default TestProvider
