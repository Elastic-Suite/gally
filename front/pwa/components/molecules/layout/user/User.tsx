import Help from '~/components/atoms/help/Help'
import UserMenu from '~/components/atoms/userMenu/UserMenu'
import { styled } from '@mui/system'
import { useContext } from 'react'
import { userContext } from '~/contexts'
import { isValidUser } from '~/../shared'

const CustomRoot = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'row',
  gap: theme.spacing(1),
  marginLeft: 'auto',
}))

function User(): JSX.Element | null {
  const user = useContext(userContext)
  const isConnected = isValidUser(user)

  if (!isConnected) return
  return (
    <CustomRoot>
      <Help />
      <UserMenu />
    </CustomRoot>
  )
}

export default User
