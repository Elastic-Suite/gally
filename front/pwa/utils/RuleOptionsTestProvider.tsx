import { ReactNode } from 'react'

import ruleEngineOperator from '~/public/mocks/rule_engine_operators.json'
import sourceFields from '~/public/mocks/source_fields.json'
import sourceFieldLabels from '~/public/mocks/source_field_labels.json'
import { RuleAttributeType } from 'shared'

import RuleOptionsProvider from '~/components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(`/source_fields/${field.id}`)
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
