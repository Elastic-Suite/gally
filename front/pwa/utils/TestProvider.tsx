import { ReactNode } from 'react'

import { schemaContext, userContext } from '~/contexts'
import { IApi, IUser } from 'shared'

import OptionsProvider from '~/components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  api: IApi
  children: ReactNode
}

const timestamp = Math.floor(Date.now() / 1000)
const expires = 8 * 60 * 60 // 8 hours

const user = {
  iat: timestamp,
  exp: timestamp + expires,
  roles: ['ROLE_ADMIN', 'ROLE_CONTRIBUTOR'],
  username: 'admin@example.com',
} as IUser

function TestProvider(props: IProps): JSX.Element {
  const { api, children } = props
  return (
    <userContext.Provider value={user}>
      <schemaContext.Provider value={api}>
        <OptionsProvider>{children}</OptionsProvider>
      </schemaContext.Provider>
    </userContext.Provider>
  )
}

export default TestProvider
