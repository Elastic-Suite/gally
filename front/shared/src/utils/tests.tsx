import { RenderOptions, render, renderHook } from '@testing-library/react'
import { ReactNode } from 'react'

import { schemaContext } from '../contexts'
import { api } from '../mocks'

export type IExtendedRenderOptions = Omit<RenderOptions, 'queries'>

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
    )
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
    )
  }
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
