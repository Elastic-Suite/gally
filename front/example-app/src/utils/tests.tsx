import { RenderOptions, render, renderHook } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { ReactNode } from 'react'
import { api } from 'gally-admin-shared'

import TestProvider from './TestProvider'

export type IExtendedRenderOptions = Omit<RenderOptions, 'queries'>

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
