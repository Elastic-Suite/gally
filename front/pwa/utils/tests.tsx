import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { ReactNode } from 'react'
import { Bundle, IUser, api } from 'shared'

import { AppStore, RootState, setupStore } from '~/store'

import AppProvider from '~/components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const timestamp = Math.floor(Date.now() / 1000)
const expires = 8 * 60 * 60 // 8 hours

const user = {
  iat: timestamp,
  exp: timestamp + expires,
  roles: ['ROLE_ADMIN', 'ROLE_CONTRIBUTOR'],
  username: 'admin@example.com',
} as IUser

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: ['https://localhost/products'],
  },
  user: {
    requestedPath: '',
    token:
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE2NTk5NTM1NDQsImV4cCI6MTY1OTk4MjM0NCwicm9sZXMiOlsiUk9MRV9BRE1JTiIsIlJPTEVfQ09OVFJJQlVUT1IiXSwidXNlcm5hbWUiOiJhZG1pbkBleGFtcGxlLmNvbSJ9.ZNE2SHmtm9WgWEMmKGMjpcfEym2ysxC_CVcLtzlpuT3jlT0Zyz4VjoQHdTav9I7s9b2QzpeDNt3Sfrbka8IHcVb_TvZle_vII-TkQGal37RtXxKk_EMMQQlzLRO3zDWYy2I0ZPXv6iMlaSWOya4egOG2s1_tzFf4MnNNm2zEQbjzAtZ_Ij9UBMlSxK19AHvHlKU9AelA-NepjUdl6f4GxYPnyIwZw3PS294Wtffj7Kzw1ZZywDljl1xN_flg70cuO0GGkUqBNHrf7zBjPmXbAZZDAefoyQfVqB2v4GzDJfO847XymvBlHtCCjWx0ThK5afyOGL2MAAZkl-AoNjMcjA',
    user,
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
