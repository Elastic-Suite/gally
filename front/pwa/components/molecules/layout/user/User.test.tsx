import { renderWithProviders } from '~/utils/tests'

import User from './User'

describe('User', () => {
  it('match snapshot not connected', () => {
    const { container } = renderWithProviders(
      <User isConnectedWithValidJwt={false} />
    )
    expect(container).toMatchSnapshot()
  })

  it('match snapshot connected', () => {
    const { container } = renderWithProviders(
      <User isConnectedWithValidJwt />
    )
    expect(container).toMatchSnapshot()
  })
})
