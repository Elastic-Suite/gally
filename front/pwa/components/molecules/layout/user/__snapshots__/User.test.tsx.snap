// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User match snapshot connected 1`] = `
<div>
  <div
    class="css-1gzl8uu"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="css-l3v3zn"
      >
        <ion-icon
          name="help-circle-outline"
          style="font-size: 18px; color: rgb(129, 135, 185);"
        />
      </div>
      <div
        class="MuiBox-root css-79elbk"
      >
        <div
          class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-199euq-MuiCollapse-root"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-1642zpf-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-xxknsj-MuiCollapse-wrapperInner"
            >
              <div
                class="css-c7ciki"
              >
                <div
                  class="css-ykyq1f"
                >
                  <a
                    class="css-ytumd6"
                    href="https://elasticsuite.zendesk.com"
                  >
                    Helpdesk
                  </a>
                  <a
                    class="css-ytumd6"
                    href="https://elastic-suite.github.io/documentation/"
                  >
                    User guide
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="css-1gbhtbv"
      >
        <ion-icon
          name="person-outline"
          style="font-size: 15px; color: rgb(129, 135, 185);"
        />
        <div
          class="css-1cn3y7o"
        >
          Admin@example.com
        </div>
        <div
          class="css-1tg3i43"
        >
          <ion-icon
            name="chevron-down"
            style="font-size: 15px; color: rgb(129, 135, 185);"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-79elbk"
      >
        <div
          class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-199euq-MuiCollapse-root"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-1642zpf-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-xxknsj-MuiCollapse-wrapperInner"
            >
              <div
                class="css-1vcaigq"
              >
                <div
                  class="css-9inyfk"
                >
                  <div
                    class="css-3lsmy"
                  >
                    Admin@example.com
                  </div>
                  <div
                    class="css-1772pc6"
                  />
                  <div
                    class="css-9vgysh"
                  >
                    Log out
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`User match snapshot not connected 1`] = `<div />`;
